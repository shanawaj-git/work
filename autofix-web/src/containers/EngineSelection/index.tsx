import Section from "@/components/Section";
import React, { useCallback, useEffect, useState } from "react";
import { Article, Button, utils } from "@albathanext/design-system";
import PageHeader from "@/components/PageHeader/PageHeader";
import { useRouter } from "next/router";
import { FormattedMessage } from "react-intl";
import BackArrow from "@/assets/arrow-back-icon.svg";
import ItemCardWithAction from "@/components/ItemCardWithAction/ItemCardWithAction";
import { LabelType } from "@/services/common";
import { getCarEngineTypesByModelIdAndYearId } from "@/services/engineTypeSelection";
import { CarEngineType } from "@/Types/CarEngine";
import { DEFAULT_ENGINE_ID } from "@/services/carOverview";
import { MILEAGE_SELECTION } from "@/app/config";

export const pageName = "engine.selection";

export interface EngineSelectionPropsType {
  en?: { labels: LabelType[] };
  ar?: { labels: LabelType[] };
}

const DEFAULT_OPTION_ID = "-1";

const EngineSelection = () => {
  const router = useRouter();
  const [engineTypes, setEngineTypes] = useState<CarEngineType[]>([]);

  const getEngineTypes = useCallback(async () => {
    const { modelId, yearId } = router.query;
    if (modelId && yearId) {
      const engineTypes = await getCarEngineTypesByModelIdAndYearId(
        modelId as string,
        yearId as string
      );

      setEngineTypes(engineTypes);
    }
  }, [router.query]);

  useEffect(() => {
    if (router.isReady) {
      getEngineTypes();
    }
  }, [router.isReady, getEngineTypes]);

  const onBackClickHandler = () => {
    router.back();
  };

  const onItemChooseHandler = (engineId: string) => {
    router.push({
      pathname: MILEAGE_SELECTION,
      query: {
        ...router.query,
        ...(engineId !== DEFAULT_ENGINE_ID && { engineId }),
        ...(engineTypes.length === 0 && { yearId: "" }),
      },
    });
  };

  return (
    <Section
      className={`flex flex-col  h-screen`}
      style={{ backgroundColor: utils.colorSelector("C500") }}
      data-testid={`${pageName}`}
    >
      <PageHeader>
        <Button
          data-testid={`${pageName}-back-button`}
          className={`mb-5 min-w-[40px] min-h-[40px] w-[40px] h-[40px]`}
          style={{
            borderColor: utils.colorSelector("b100"),
            backgroundColor: utils.colorSelector("P700"),
          }}
          onClick={onBackClickHandler}
        >
          <img src={BackArrow} alt="BackArrow" />
        </Button>
        <Article
          title={<FormattedMessage id={`${pageName}.header.title`} />}
          titleProps={{
            color: utils.colorSelector("P200"),
            className: "text-xl whitespace-pre-line !mt-4",
          }}
        />
      </PageHeader>

      <div className="m-[24px]">
        {engineTypes.map((engineType: CarEngineType) => (
          <ItemCardWithAction
            key={engineType.id}
            itemId={engineType.id}
            itemText={engineType.name}
            onClick={() => onItemChooseHandler(engineType.id)}
          />
        ))}
        <ItemCardWithAction
          key={DEFAULT_OPTION_ID}
          itemId={DEFAULT_OPTION_ID}
          itemText={<FormattedMessage id={`${pageName}.option.dontknow`} />}
          onClick={() => onItemChooseHandler(DEFAULT_OPTION_ID)}
        />
      </div>
    </Section>
  );
};

export default EngineSelection;
