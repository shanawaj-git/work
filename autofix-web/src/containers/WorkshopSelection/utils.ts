import {
  BoundPayloadType,
  GetDirections,
  LatLngType,
  MapObjType,
} from "./interfaces";

export const markerIcon = (icon: string) => ({
  scaledSize: {
    height: 24,
    width: 24,
  },
  url: icon,
});

export const getBoundLocation = ({
  map,
  onBoundChange,
}: {
  map: MapObjType;
  onBoundChange: (e: BoundPayloadType) => void;
}) => {
  const bounds = map.getBounds();
  const southWest = getSouthWestCoordinates(bounds);
  const northEast = getNorthEastCoordinates(bounds);
  const result = buildBoundPayload(northEast, southWest);
  onBoundChange(result);
};

const getSouthWestCoordinates = (bounds: GetDirections): LatLngType => ({
  lat: bounds.getSouthWest().lat(),
  lng: bounds.getSouthWest().lng(),
});

const getNorthEastCoordinates = (bounds: GetDirections): LatLngType => ({
  lat: bounds.getNorthEast().lat(),
  lng: bounds.getNorthEast().lng(),
});

const buildBoundPayload = (
  northEast: LatLngType,
  southWest: LatLngType
): BoundPayloadType => ({
  locationBoxBounds: {
    bottomLeft: {
      type: "Point",
      coordinates: [southWest.lng, southWest.lat],
    },
    topRight: { type: "Point", coordinates: [northEast.lng, northEast.lat] },
  },
});
