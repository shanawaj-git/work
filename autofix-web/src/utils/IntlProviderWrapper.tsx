import React, { useMemo } from "react";
import { getTranslation, TranslationsType } from "../translations";
import { IntlProvider } from "react-intl";
import { flatData, I18NLocale } from ".";

export const withWrapper = (component: JSX.Element, pageProps = { en: {} }) => (
  <IntlProviderWrapper locale="en" pageProps={pageProps}>
    {component}
  </IntlProviderWrapper>
);

export const IntlProviderWrapper = ({
  children,
  pageProps,
  locale,
}: {
  children: JSX.Element;
  pageProps: TranslationsType;
  locale: I18NLocale;
}) => {
  const messages = useMemo(() => {
    return {
      ...getTranslation(locale),
      ...flatData(pageProps[locale] || {}),
    };
  }, [locale, pageProps]);

  return (
    <IntlProvider locale={`${locale}`} messages={messages}>
      {children}
    </IntlProvider>
  );
};
