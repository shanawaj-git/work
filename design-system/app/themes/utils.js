import { getTheme } from './getTheme';

const BOX_SHADOW_MAPPING = {
  level1: '0px 3px 10px rgba(0, 0, 0, 0.07)',
  level2: '0px 6px 10px rgba(0, 0, 0, 0.07)',
  level3: '0px 10px 20px rgba(0, 0, 0,  0.07)',
  level4: '0px 14px 28px rgba(0, 0, 0, 0.1)',
  level5: '0px 19px 38px rgba(0, 0, 0, 0.15)',
  focus: `0px 0px 10px ${getTheme().theme.color.d100}`,
};

const boxShadowSelector = level => BOX_SHADOW_MAPPING[level];
const colorSelector = color => getTheme().theme.color[color.toLowerCase()];
export { boxShadowSelector, colorSelector };
