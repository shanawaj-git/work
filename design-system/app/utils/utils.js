import { v4 } from 'uuid';
import _debounce from 'lodash/debounce';
/**
 * Helper: Convert Array of coordinates to an object of cordinates
 * @param {*} cordinates : Array [lat,lng]
 * @returns
 */
export const transformedCordinates = cordinates => {
  const [lat, lng] = cordinates;
  return {
    lat,
    lng,
  };
};

export const readableCordinates = ({ lat, lng }) => ({
  lat: lat(),
  lng: lng(),
});

export const uuid = () => v4();

const MAP_ADDRESS_COMPONENT_TYPE = {
  LOCALITY: 'locality',
  COUNTRY: 'country',
};

export const getCityAndCountry = address => {
  let city;
  let country;
  // eslint-disable-next-line no-unused-expressions
  address?.forEach(component => {
    const { types } = component;
    if (types.indexOf(MAP_ADDRESS_COMPONENT_TYPE.LOCALITY) > -1) {
      city = component.long_name;
    }

    if (types.indexOf(MAP_ADDRESS_COMPONENT_TYPE.COUNTRY) > -1) {
      country = component.long_name;
    }
  });
  return {
    city,
    country,
  };
};

export const debounce = (func, wait) => _debounce(func, wait);

export const canBeRenderedInTypographyComponent = obj =>
  typeof obj === 'string' ||
  obj?.type?.displayName === 'MemoizedFormattedMessage';
