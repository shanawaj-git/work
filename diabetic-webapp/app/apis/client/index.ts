import { APIClientOptions } from 'apis/types';
import axios from 'axios';
import { addInterceptors } from './addInterceptors';
import * as environmentConfig from '../../configs/environmentConfig';

let httpClient;
// eslint-disable-next-line no-unused-vars
const configureAPIClient = (options: APIClientOptions) => {
  // TODO: intercept response and add callback support for login, auth failed
  httpClient = axios.create({
    baseURL:
      environmentConfig.getEnvironmentConfig().API_BASE_URL ||
      'http://localhost:1337',
    timeout: 30000, // 30 seconds
    // TODO: pass Authorization once login is implemented
  });
  addInterceptors(httpClient, options);
  return httpClient;
};

const getAPIClientInstance = () => {
  if (!httpClient) {
    throw new Error('API Client is not intialized');
  }
  return httpClient;
};

export { configureAPIClient, getAPIClientInstance, APIClientOptions };
