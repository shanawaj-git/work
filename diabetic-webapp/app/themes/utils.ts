
import { diabeticTheme } from './theme';

const BOX_SHADOW_MAPPING = {
  level1: '0px 3px 10px rgba(0, 0, 0, 0.07)',
  level2: '0px 6px 10px rgba(0, 0, 0, 0.07)',
  level3: '0px 10px 20px rgba(0, 0, 0,  0.07)',
  level4: '0px 14px 28px rgba(0, 0, 0, 0.1)',
  level5: '0px 19px 38px rgba(0, 0, 0, 0.15)',
};

const boxShadowSelector = (level: string): string => BOX_SHADOW_MAPPING[level];

const colorSelector = (color: string): string =>
  diabeticTheme.color[color.toLowerCase()];

export { boxShadowSelector, colorSelector };
