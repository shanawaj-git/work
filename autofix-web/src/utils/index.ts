import { itemType } from "@/containers/RecommendedMaintenance";
import {
  AssetType,
  fetchCommonData,
  FooterType,
  getAssets,
  getAssetsData,
  getLabelData,
  getLabels,
  HomeActionType,
  LabelType,
  ServiceType,
} from "@/services/common";
import { TestimonialType } from "@/services/home/testimonial";

export const utils = "UTILS";
const GRAPHQL_CLIENT_READ_TIMEOUT = 5000;

interface getCommonPageContentParams {
  locale: I18NLocale;
  pageName: string;
}

export interface HomeContentType {
  labels?: LabelType[];
  assets?: AssetType[];
  services?: ServiceType[];
  actions?: HomeActionType[];
  testimonials?: TestimonialType[];
  footer?: FooterType;
}

export type I18NLocale = "en" | "ar";

/**
 * Will Return a Data object with Labels and Assets Regarding the page from Strapi
 * @param  locale , pageName
 * @returns Object with Labels and Assets
 */

export const getCommonPageContent = async ({
  locale,
  pageName,
}: getCommonPageContentParams): Promise<HomeContentType> => {
  let data: HomeContentType = {};
  return new Promise(async (resolve, reject) => {
    try {
      const commonData = await (
        await fetchCommonData(locale as I18NLocale, pageName)
      ).data;
      const assets = getAssets(commonData);
      const assetsData = getAssetsData(assets);
      const labels = getLabelData(getLabels(commonData));
      data = { assets: assetsData, labels };
      resolve(data);
    } catch (error) {
      resolve(data);
    }
  });
};

interface rejectEarlyType<T> {
  fn: () => Promise<T>;
  def: T;
  timeout?: number;
}

export const rejectEarly = <T>({
  fn,
  def,
  timeout = GRAPHQL_CLIENT_READ_TIMEOUT,
}: rejectEarlyType<T>): Promise<T> => {
  return new Promise(async (resolve, _reject) => {
    setTimeout(() => {
      resolve(def);
    }, timeout);
    try {
      const out: T = await fn();
      resolve(out);
    } catch (error) {
      resolve(def);
    }
  });
};

export const validateEmail = (email: string) =>
  /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email);

export const flatData = (obj: { [key: string]: {} }) =>
  Object.values(obj).reduce((acc: {}, curr: {}) => {
    return { ...acc, ...curr };
  }, {});

export const buildImageUrl = (path: string): string =>
  process.env.NEXT_PUBLIC_STRAPI_SERVICE?.concat(path) || "";

export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string
) =>
  array.reduce((acc, value, index, array) => {
    (acc[predicate(value, index, array)] ||= []).push(value);
    return acc;
  }, {} as { [key: string]: T[] });

export const includesCaseInsensitive = (
  string: string,
  substring: string
): Boolean => {
  return string.toLowerCase().includes(substring.toLowerCase());
};
export const getTotalForMax = (data: itemType[]) => {
  return data.reduce((acc: number, curr: { cost: string }) => {
    let value = acc;
    value += +parseFloat(
      (curr.cost.match(/[0-9]+/gm) || ["0"]).splice(-1)[0]
    ).toFixed(2);
    return value;
  }, 0);
};

export const getCurrency = (data: itemType[]) => {
  return (data[0].cost.match(/[A-Z]+/gm) || ["USD"]).splice(-1)[0];
};

export const getCurrentLocation = () => {
  return new Promise((resolve: (e: GeolocationCoordinates) => void, reject) => {
    if (!navigator) throw new Error("Navigator doesn't exist");
    if (!navigator.geolocation) throw new Error("Geolocation doesn't exist");
    navigator.geolocation.getCurrentPosition(
      (position: GeolocationPosition) => {
        return resolve(position.coords);
      },
      reject,
      {
        maximumAge: 5000,
        enableHighAccuracy: true,
      }
    );
  });
};

interface getCurrentLiveLocationType {
  callBack: (e: GeolocationPosition) => void;
  onError: (e: GeolocationPositionError | Error) => void;
}

export const getCurrentLiveLocation = ({
  callBack,
  onError,
}: getCurrentLiveLocationType) => {
  if (!navigator) throw new Error("Navigator doesn't exist ");
  if (!navigator.geolocation) throw new Error("Geolocation doesn't exist ");
  navigator.geolocation.watchPosition((position: GeolocationPosition) => {
    callBack(position);
  }, onError);
};
