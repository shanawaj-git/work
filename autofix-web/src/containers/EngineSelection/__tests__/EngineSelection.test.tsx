import React from "react";
import { act, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import EngineSelection, { pageName } from "..";
import { withWrapper } from "@/utils/IntlProviderWrapper";
import { useRouter } from "next/router";
import * as mockCarsQueryResponse from "./CarsQueryResponse.json";
import client from "@/apollo/index";

const asyncRenderComponent = async () => {
  let tree;
  await act(async () => {
    tree = render(withWrapper(<EngineSelection />));
  });
  return tree;
};
describe("<EngineSelection />", () => {
  const router = useRouter();
  router.isReady = true;
  router.query = { modelId: "1", yearId: "1" };
  beforeEach(() => {
    client.query.mockResolvedValue(mockCarsQueryResponse);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("it should mount", async () => {
    await asyncRenderComponent();
    const engineSelection = screen.getByTestId(pageName);
    expect(engineSelection).toBeInTheDocument();
  });

  test("it should mount and match to the snapshot", async () => {
    const tree = await asyncRenderComponent();
    expect(tree.container.innerHTML).toMatchSnapshot();
  });

  test(`it should have title 'Select your engine type'`, async () => {
    const tree = await asyncRenderComponent();
    const titleElement = tree.getByTestId("article-component-title");
    const titleElementSpan = titleElement.getElementsByTagName("span");
    expect(titleElementSpan[0].innerHTML).toBe(" Select your engine type");
  });

  test(`it should have default engineType option called I don't know`, async () => {
    const tree = await asyncRenderComponent();
    const defaultItem = tree.getByText(/I don't know/);
    expect(defaultItem).toBeInTheDocument();
  });

  test(`when engine Types query resolved 2 items, Use should see 3 options including I don't know'`, async () => {
    const tree = await asyncRenderComponent();
    const items = tree.getAllByTestId(/itemCardWithAction-/);
    const defaultItem = tree.getByText(/I don't know/);
    expect(defaultItem).toBeInTheDocument();
    expect(items.length).toBe(3);
  });

  test(`when user chooses I don't know, it should redirect to mileage-selection page with engineId -1`, async () => {
    const expectedRouterOptions = {
      pathname: "/mileage-selection",
      query: {
        modelId: "1",
        yearId: "1",
      },
    };
    const tree = await asyncRenderComponent();
    const defaultItem = tree.getByText(/I don't know/);
    fireEvent(defaultItem, new MouseEvent("click", { bubbles: true }));
    expect(router.push).toBeCalledTimes(1);
    expect(router.push).toBeCalledWith(expectedRouterOptions);
  });

  test(`when user chooses Engine Type, it should redirect to mileage-selection page`, async () => {
    const expectedRouterOptions = {
      pathname: "/mileage-selection",
      query: {
        engineId: "1",
        modelId: "1",
        yearId: "1",
      },
    };
    const tree = await asyncRenderComponent();
    const defaultItem = tree.getByText(/1.6 VVT i/);
    fireEvent(defaultItem, new MouseEvent("click", { bubbles: true }));
    expect(router.push).toBeCalledWith(expectedRouterOptions);
  });

  test("when user clicks back button, should redirect to previous route and it calls only once", async () => {
    const tree = await asyncRenderComponent();
    const button = tree.getByTestId(`${pageName}-back-button`);
    fireEvent(button, new MouseEvent("click", { bubbles: true }));
    expect(router.back).toBeCalledTimes(1);
  });

  test(`when engine Types query resolved empty, It should render only one option named 'I don't know'`, async () => {
    client.query.mockResolvedValue({
      data: {
        cars: {
          data: [],
        },
      },
    });

    const expectedRouterOptions = {
      pathname: "/mileage-selection",
      query: {
        modelId: "1",
        yearId: "",
      },
    };

    const tree = await asyncRenderComponent();
    const items = tree.getAllByTestId(/itemCardWithAction-/);
    const defaultItem = tree.getByText(/I don't know/);
    expect(defaultItem).toBeInTheDocument();
    expect(items.length).toBe(1);
    fireEvent(defaultItem, new MouseEvent("click", { bubbles: true }));
    expect(router.push).toBeCalledWith(expectedRouterOptions);
  });
});
