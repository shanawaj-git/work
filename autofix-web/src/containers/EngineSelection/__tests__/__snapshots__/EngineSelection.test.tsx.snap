// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EngineSelection /> it should mount and match to the snapshot 1`] = `"<div data-testid=\\"engine.selection\\" class=\\"min-h-screen flex flex-col  h-screen\\"><div data-testid=\\"PageHeader\\" class=\\"min-h-screen p-6 \\" style=\\"min-height: fit-content;\\"><button class=\\"root StyledButton-sc-1apep4b-0 bXswYh mb-5 min-w-[40px] min-h-[40px] w-[40px] h-[40px] MuiButton-root text MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root\\" tabindex=\\"0\\" type=\\"button\\" data-testid=\\"engine.selection-back-button\\" style=\\"border-color: #c4f8cc;\\"><div style=\\"visibility: visible;\\"><img src=\\"arrow-back-icon.svg\\" alt=\\"BackArrow\\"></div><span class=\\"MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root\\"></span></button><div data-testid=\\"article-component\\"><p class=\\"MuiTypography-root MuiTypography-body1 text-xl whitespace-pre-line !mt-4 css-1uw3dvl-MuiTypography-root\\" data-testid=\\"article-component-title\\"><span> Select your engine type</span></p><div color=\\"#FCF7F7\\" class=\\"text-xl whitespace-pre-line !mt-4\\"></div></div></div><div class=\\"m-[24px]\\"><div data-testid=\\"itemCardWithAction-1\\" class=\\"flex p-5 mt-3 items-center justify-between rounded border cursor-pointer border-solid border-[#00000014]\\" style=\\"gap: 10px; background-color: rgb(255, 255, 255);\\"><p class=\\"font-normal font-[500] text-base leading-6\\">1.6 VVT i</p><div><img class=\\"min-h-3.5 min-w-3 left-1.25 top-4.5 rounded-none\\" src=\\"arrow-forward-icon.svg\\" alt=\\"forward action icon\\"></div></div><div data-testid=\\"itemCardWithAction-2\\" class=\\"flex p-5 mt-3 items-center justify-between rounded border cursor-pointer border-solid border-[#00000014]\\" style=\\"gap: 10px; background-color: rgb(255, 255, 255);\\"><p class=\\"font-normal font-[500] text-base leading-6\\">2.0 VVT i</p><div><img class=\\"min-h-3.5 min-w-3 left-1.25 top-4.5 rounded-none\\" src=\\"arrow-forward-icon.svg\\" alt=\\"forward action icon\\"></div></div><div data-testid=\\"itemCardWithAction--1\\" class=\\"flex p-5 mt-3 items-center justify-between rounded border cursor-pointer border-solid border-[#00000014]\\" style=\\"gap: 10px; background-color: rgb(255, 255, 255);\\"><p class=\\"font-normal font-[500] text-base leading-6\\">I don't know</p><div><img class=\\"min-h-3.5 min-w-3 left-1.25 top-4.5 rounded-none\\" src=\\"arrow-forward-icon.svg\\" alt=\\"forward action icon\\"></div></div></div></div>"`;
